src/tls/server.c

```diff
@@ -309,6 +309,44 @@ handle_accept (void)
   server.connections = con;
 }
 
+/**
+ * verify_peer_certificate: Custom client certificate validation function
+ *
+ * cockpit-tls ignores CA/trusted owner and leaves that to e. g. sssd. But
+ * validate the other properties such as expiry, unsafe algorithms, etc.
+ * This combination cannot be done with gnutls_session_set_verify_cert().
+ */
+static int
+verify_peer_certificate (gnutls_session_t session)
+{
+  int ret;
+  unsigned status;
+
+  TLS_RETRY_BLOCK (ret, gnutls_certificate_verify_peers2 (session, &status));
+  if (ret >= 0)
+    {
+      /* ignore CA/trusted owner and leave that to e. g. sssd */
+      status &= ~(GNUTLS_CERT_INVALID | GNUTLS_CERT_SIGNER_NOT_FOUND | GNUTLS_CERT_SIGNER_NOT_CA);
+      if (status != 0)
+        {
+          gnutls_datum_t msg;
+          ret = gnutls_certificate_verification_status_print (status, gnutls_certificate_type_get (session), &msg, 0);
+          if (ret != GNUTLS_E_SUCCESS)
+            errx (1, "Failed to print verification status: %s", gnutls_strerror (ret));
+          warnx ("Invalid TLS peer certificate: %s", msg.data);
+          gnutls_free (msg.data);
+          return GNUTLS_E_CERTIFICATE_VERIFICATION_ERROR;
+        }
+    }
+  else if (ret != GNUTLS_E_NO_CERTIFICATE_FOUND)
+    {
+      warnx ("Verifying TLS peer failed: %s", gnutls_strerror (ret));
+      return ret;
+    }
+
+  return 0;
```

Comment:
`GNUTLS_E_SUCCESS`?

Reply:
I'm a bit torn on that. The manpage says "The callback function should return 0 for the handshake to continue or non-zero to terminate.", but as it clearly accepts GNUTLS_E_* error codes, this may seem cleaner.

Reply:
Updated.